src/Common/src/Interop/OSX/Interop.libproc.cs

```diff
@@ -0,0 +1,523 @@
+            internal ulong       ri_cpu_time_qos_utility;
+            internal ulong       ri_cpu_time_qos_legacy;
+            internal ulong       ri_cpu_time_qos_user_initiated;
+            internal ulong       ri_cpu_time_qos_user_interactive;
+            internal ulong       ri_billed_system_time;
+            internal ulong       ri_serviced_system_time;
+        }
+
+        // From proc_info.h
+        [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Ansi)]
+        internal struct proc_taskallinfo
+        {
+            internal proc_bsdinfo    pbsd;
+            internal proc_taskinfo   ptinfo;
+        }
+
+        // From proc_info.h
+        [StructLayout(LayoutKind.Sequential)]
+        internal unsafe struct proc_threadinfo
+        {
+            internal ulong pth_user_time;
+            internal ulong pth_system_time;
+            internal int pth_cpu_usage;
+            internal int pth_policy;
+            internal int pth_run_state;
+            internal int pth_flags;
+            internal int pth_sleep_time;
+            internal int pth_curpri;
+            internal int pth_priority;
+            internal int pth_maxpriority;
+            internal fixed byte pth_name[MAXTHREADNAMESIZE];
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct proc_fdinfo
+        {
+            internal int proc_fd;
+            internal uint proc_fdtype;
+        }
+
+        /// <summary>
+        /// Queries the OS for the PIDs for all running processes
+        /// </summary>
+        /// <param name="buffer">A pointer to the memory block where the PID array will start</param>
+        /// <param name="buffersize">The length of the block of memory allocated for the PID array</param>
+        /// <returns>Returns the number of elements (PIDs) in the buffer</returns>
+        [DllImport(Interop.Libraries.libproc, SetLastError = true)]
+        private static extern int proc_listallpids(
+            IntPtr buffer, 
+            int buffersize);
+
+        /// <summary>
+        /// Queries the OS for the list of all running processes and returns the PID for each
+        /// </summary>
+        /// <returns>Returns a list of PIDs corresponding to all running processes</returns>
+        internal static int[] proc_listallpids()
+        {
+            // Get the number of processes currently running to know how much data to allocate
+            int numProcesses = proc_listallpids(IntPtr.Zero, 0);
+            if (numProcesses <= 0)
+            {
+                throw new System.Runtime.InteropServices.COMException(SR.CantGetAllPids, Marshal.GetLastWin32Error());
+            }
+
+            // Allocate the correct amount of memory, plus 1/4 more just in case more processes have spawned between
+            // when we asked and now (very possible), and then get the PID list
+            int bufferSize = Convert.ToInt32(numProcesses * Marshal.SizeOf<uint>() * (numProcesses * .25));
```

Comment:
This formula looks wrong.  From the comment, I'd have expected:
```
(int)((numProcesses * 1.25) * Marshal.SizeOf<uint>());
```
Instead the formula appears to be computing enough slots for .25numProcesses^2 elements.

Separately, what happens if there are more processes than space is provided for?  Should this be in a loop in case more than a 1/4 more processes are created in that window?